results-processor/processor.py

```diff
@@ -18,200 +21,282 @@
+        return self
+
+    def __exit__(self, *args):
+        shutil.rmtree(self._temp_dir)
+        shutil.rmtree(self._upload_dir)
+
+    def _get_auth(self):
+        user = self._datastore.get(
+            self._datastore.key('Uploader', self.USERNAME))
+        return (user['Username'], user['Password'])
+
+    @property
+    def raw_results_gs_url(self):
+        return 'gs://{}/{}/report.json'.format(
+            config.raw_results_bucket(), self.report.sha_product_path)
+
+    @property
+    def raw_results_url(self):
+        return gsutil.gs_to_public_url(self.raw_results_gs_url)
+
+    @property
+    def results_gs_url(self):
+        return 'gs://{}/{}'.format(
+            config.results_bucket(), self.report.sha_summary_path)
+
+    @property
+    def results_url(self):
+        return gsutil.gs_to_public_url(self.results_gs_url),
+
+    def check_existing_run(self):
+        """Returns true if an existing run already has raw_results_url.
+
+        This is used to abort early if the result already exists in Datastore.
+        It is safe because raw_results_url contains both the full revision &
+        checksum of the report content, unique enough to use as a UID.
+
+        Datastore does not support a query-and-put transaction, so this is
+        only a best effort to avoid duplicate runs.
+        """
+        q = self._datastore.query(kind='TestRun')
+        q.add_filter('RawResultsURL', '=', self.raw_results_url)
+        q.keys_only()
+        run = list(q.fetch(limit=1))
+        return len(run) > 0
+
+    @staticmethod
+    def known_extension(path):
+        """Returns the extension of the path if known, otherwise None."""
+        EXT = ('.json.gz', '.txt.gz', '.gz', '.zip', '.json', '.txt')
+        for e in EXT:
+            if path.endswith(e):
+                return e
+        return None
+
+    def _download_gcs(self, gcs):
+        assert gcs.startswith('gs://')
+        ext = self.known_extension(gcs)
+        fd, path = tempfile.mkstemp(suffix=ext, dir=self._temp_dir)
+        os.close(fd)
+        # gsutil will log itself.
+        gsutil.copy(gcs, path)
+        return path
+
+    def _download_http(self, url):
+        assert url.startswith('http://') or url.startswith('https://')
+        _log.debug("Downloading %s", url)
+        r = requests.get(url, stream=True)
+        ext = (self.known_extension(r.header['Content-Disposition']) or
+               self.known_extension(url))
+        fd, path = tempfile.mkstemp(suffix=ext, dir=self._temp_dir)
+        with os.fdopen(fd, mode='wb') as f:
+            for chunk in r.iter_content(chunk_size=512*1024):
+                f.write(chunk)
+        os.close(fd)
+        return path
+
+    def _download_single(self, uri):
+        if uri.startswith('gs://'):
+            return self._download_gcs(uri)
+        return self._download_http(uri)
+
+    def _download_azure(self, azure_url):
+        artifact = self._download_http(azure_url)
```

Comment:
If this fails I guess the task will be retried later?

Reply:
Yes. It will throw and the server will return 500, which will cause TaskQueue to retry.